debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/infinispan/InfinispanLogMinerTransactionCache.java

```diff
@@ -0,0 +1,247 @@
+                    return event != null && event.getRowId().equals(rowId);
+                })
+                .findFirst()
+                .map(entry -> {
+                    eventCache.remove(entry.getValue());
+                    eventIds.remove(entry.getKey());
+                    return true;
+                })
+                .orElse(false);
+    }
+
+    @Override
+    public boolean containsTransactionEvent(InfinispanTransaction transaction, int eventKey) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId()).contains(eventKey);
+    }
+
+    @Override
+    public int getTransactionEventCount(InfinispanTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId()).size();
+    }
+
+    @Override
+    public int getTransactionEvents() {
+        return eventIdsByTransactionId.values().stream().mapToInt(Set::size).sum();
+    }
+
+    @Override
+    public void clear() {
+        // Wait for all writes before we clear the cache.
+        waitForAllWrites();
+
+        transactionCache.clear();
+        eventCache.clear();
+        eventIdsByTransactionId.clear();
+    }
+
+    @Override
+    public void resetTransactionToStart(InfinispanTransaction transaction) {
+        super.resetTransactionToStart(transaction);
+        syncTransaction(transaction);
+    }
+
+    @Override
+    public void syncTransaction(InfinispanTransaction transaction) {
+        // todo:
+        // Perhaps we can look at pulling number of events out of Transaction and let that
+        // be managed in the cache's heap, in which case we can avoid this put.
+
+        // Necessary to synchronize state
+        putAsync(transactionCache, transaction.getTransactionId(), transaction);
+    }
+
+    @Override
+    public void close() throws Exception {
+        // Makes sure during shutdown that all pending writes are flushed
+        waitForAllWrites();
+    }
+
+    private void primeEventCountsByTransactionId() {
+        // Primes the heap-based cache if the Infinispan disk caches contained data on start-up
+        transactions(stream -> stream.forEach(transaction -> {
+            try (Stream<String> keyStream = eventCache.keySet().stream()) {
+                eventIdsByTransactionId.put(
+                        transaction.getTransactionId(),
+                        keyStream.map(k -> k.split("-", 2))
+                                .filter(parts -> parts.length == 2)
+                                .filter(parts -> parts[0].equals(transaction.getTransactionId()))
+                                .map(parts -> Integer.parseInt(parts[1]))
+                                .collect(Collectors.toCollection(TreeSet::new)));
+            }
+        }));
+    }
+
+    private void waitForAllWrites() {
+        for (CompletableFuture<?> future : pendingWrites) {
+            try {
+                future.get();
+            }
+            catch (Exception e) {
```

Comment:
If this is an InterruptedException, seems like maybe the loop should stop.  Also, would want to call Thread.currentThread().interrupt again.